range-diff presentation

gerhardol opened the pull request:
Fixes #8448 

Based on #8471 . Except for one line, this is independent of #8471, but it is meaningless as nothing is changed. However, changes in #8471 should result in minimal changes this PR. This is therefore not a draft.

## Proposed changes

Present range-diff output as a special diff variant with highlighting etc.
- ViewMode for range-diff, to limit menu actions (copy patches, apply patches)
- Next/previous change moves to the commit-pair header. The important with this feature is not really to see the patches
- Set syntax highlight to first found filename
- Highlight for content:
  - Commit-pair "headers" - AuthoredHighlight
  - Filename or patch start (very random contents...) - DiffSection
  - Lines only changed in last selected - DiffAddedExtra/DiffRemovedExtra
  - Lines only changed in first or same change in both - DiffAdded/DiffRemoved

Some notes of what is not done:
 * Parsing the output, separating the commit pairs (the files in the pairs are much too instable right now). This would have been a part of #8471. The overview is better, but this will require another command when selecting. This is not 
 * "Patch" (including new/old) copying. I actually implemented but removed this, as it is unclear what new/old/patch really is here
 * Use Git "dual coloring" information. Hard to handle, not giving much better information.
 * New theme colors to display output better
 * Some way to explain what is seen - there is no good description in the git-range-diff command either...

This will be used for an alternative implementation of #7825 too.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/95000775-1640b380-05c4-11eb-93c0-c01f72075e67.png)

### After

![image](https://user-images.githubusercontent.com/6248932/95000623-d5946a80-05c2-11eb-8bdf-411e9041a1d1.png)

![image](https://user-images.githubusercontent.com/6248932/95000612-b8f83280-05c2-11eb-850e-28ed6099625f.png)

![image](https://user-images.githubusercontent.com/6248932/95000719-76832580-05c3-11eb-9fa6-2bdb76265865.png)

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Rebased on master and updated #8471 (code review comments, no functionality changes) without changing this commit use this PR to see this easily. It is often more important to see the commit pair matching than what is actually changed.

![image](https://user-images.githubusercontent.com/6248932/95021129-7e000880-066f-11eb-9103-cf1701565719.png)

gerhardol commented:
Rebased, added a possibility to see commit diff only, from https://github.com/gitextensions/gitextensions/pull/8471#discussion_r501535833

gerhardol commented:
Added a fixup to the possibility to see commit ranges only
Squashed, rebased on #8471

Previous commits in https://github.com/gerhardol/gitextensions/tree/feature/i8448-presentation-1

mstv commented:
The `Show entire file` button could be re-used to show/hide the range-diff details, perhaps rather with a dedicated setting for range-diffs indepent from that for patches. (follow-up PR?)

---

The not applicable context menu items should be hidden.

gerhardol commented:
> 
> 
> The `Show entire file` button could be re-used to show/hide the range-diff details, perhaps rather with a dedicated setting for range-diffs indepent from that for patches. (follow-up PR?)
> 
> The not applicable context menu items should be hidden.

I considered this, but it has a meaning for range-diff too, show the maximum diff possible
Also the kind of expected behavior is the opposite of for normal diffs, as you would need to use "show entire file" to see any diffs at all, which you have to switch back for normal diffs.

mstv commented:
> which you have to switch back for normal diffs

That's why "perhaps rather with a dedicated setting for range-diffs indepent from that for patches"

gerhardol commented:
> That's why "perhaps rather with a dedicated setting for range-diffs indepent from that for patches"

I would like to keep this as a possible improvement for the future for now, to not change too much
I hope someone has a better idea for the presentation, for instance that "async diff calc" is done to parse the output and present it better.
Or that Git range-diff diff part is not just giving approximate diff...

mstv commented on GitUI/Editor/FileViewerInternal.cs:
```diff
@@ -233,12 +233,17 @@ public string GetText()
         public bool? ShowLineNumbers { get; set; }
 
         public void SetText(string text, Action openWithDifftool, bool isDiff = false)
+        {
+            SetText(text, openWithDifftool, isDiff, false);
+        }
+
+        public void SetText(string text, Action openWithDifftool, bool isDiff, bool isRangeDiff)
```
```suggestion
        public void SetText(string text, Action openWithDifftool, bool isDiff = false, bool isRangeDiff = false)
```

gerhardol replied:
public interface...

mstv commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -332,14 +335,17 @@ public void EnableScrollBars(bool enable)
                 { ShowEntireFile, "--inter-hunk-context=9000 --unified=9000", $"--unified={NumberOfContextLines}" },
+
+                // Handle zero context as showing no file changes, to get the summary only
+                { isRangeDiff && NumberOfContextLines == 0, "--no-patch " },
```
Not urgent because the buttons `Next` /  `Prev` jump between commits. 👍

gerhardol commented on GitUI/Editor/Diff/RangeDiffHighlightService.cs:
```diff
@@ -0,0 +1,79 @@
+
+        protected override int GetDiffContentOffset()
+        {
+            return 6;
```
Added a comment

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -1345,11 +1364,14 @@ private void CopyToolStripMenuItemClick(object sender, EventArgs e)
-                var specials = isCombinedDiff ? new[] { "  ", "++", "+ ", " +", "--", "- ", " -" }
-                    : new[] { " ", "-", "+" };
+                var specials = _viewMode == ViewMode.RangeDiff
+                    ? new[] { "      ", "    ++", "    + ", "     +", "    --", "    - ", "     -", "    +-", "    -+", "    " }
```
I have assumed so - similar to before

RussKie replied:
I don't think it would. I think we'd be allocating a new array every time. I'd define a static array at the class level.

Ditto below

gerhardol replied:
Allocated static arrays (if the performance can be measured, it is for existing diffs)

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -470,6 +476,11 @@ public async Task ViewFixedPatchAsync(string fileName, string text, Action openW
 
+        public async Task ViewRangeDiffAsync(string fileName, string text)
+        {
+            await ViewPrivateAsync(fileName, text, null, ViewMode.RangeDiff);
```
Both!

gerhardol commented on GitUI/Editor/FileViewerInternal.cs:
```diff
@@ -248,11 +253,16 @@ public void SetText(string text, Action openWithDifftool, bool isDiff = false)
+                ? RangeDiffHighlightService.Instance
+                : DiffHighlightService.IsCombinedDiff(text)
                     ? CombinedDiffHighlightService.Instance
                     : DiffHighlightService.Instance;
```
_Format selection_ does not handle this...

gerhardol commented:
rebased on #8471, review comments

mstv commented:
Final request: Strip down the FileStatusList context menu for range-diff, please. Perhaps in #8471 already.

gerhardol commented:
> Final request: Strip down the FileStatusList context menu for range-diff, please. Perhaps in #8471 already.

OK, for #8471 
(and you are OK to change the Final for this PR too..)

mstv commented:
> (and you are OK to change the Final for this PR too..)

For me, yes because I strongly prefer development tools in English. But I just remembered that the hint with the command should be made translatable.

gerhardol commented:
> But I just remembered that the hint with the command should be made translatable.

That is also #8471 
I would prefer to not bother translators with this (there are some similar for cannot displaying images etc I believe too).

mstv commented on GitUI/UserControls/FileStatusDiffCalculator.cs:
```diff
@@ -0,0 +1,230 @@
+            // Add rangeDiff as a separate group (range is not the same as diff with artificial commits)
+            List<GitItemStatus> statuses = new List<GitItemStatus> { new GitItemStatus { Name = Strings.DiffRange, IsRangeDiff = true } };
+
+            var desc = Strings.DiffRange + ": " + GetDescriptionForRevision(describeRevision, firstRevHead) + " -> " +
+                       GetDescriptionForRevision(describeRevision, selectedRevHead);
```
Make use of `$"{var}"`?

gerhardol replied:
Too late, that was #8471!
To avoid unexpected reviews, I have rebased on master, now that #8471 is merged.
(Could of course change that here too, but I find that the line get too long for $"" to be useful. Could add two temp variables though.)

mstv replied:
That's why as question not as request. This shouldn't matter because not used extensively. So rather keep it as it is.

mstv reviewed (approved):
OK for me, I was waiting for votes regarding the minor open points.

gerhardol commented:
Allocated static arrays. (Had first planned to ignore this in this PR).
To get rid of one of them I refactored a little and fixed that previous change for combined diff did not work (will squash that, even if unrelated to the PR change).
Will appreciate if @mstv takes a quick look, then I will squash and set to merge some days later.

> But I just remembered that the hint with the command should be made translatable.

Still skipping this....

RussKie commented:
👍

On Wed, 21 Oct. 2020, 9:54 am Gerhard Olsson, <notifications@github.com>
wrote:

> Allocated static arrays. (Had first planned to ignore this in this PR).
> To get rid of one of them I refactored a little and fixed that previous
> change for combined diff did not work (will squash that, even if unrelated
> to the PR change).
> Will appreciate if @mstv <https://github.com/mstv> takes a quick look,
> then I will squash and set to merge some days later.
>
> But I just remembered that the hint with the command should be made
> translatable.
>
> Still skipping this....
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8516#issuecomment-713185266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXSE6VOO5HDUYW55OSLSLYIJLANCNFSM4SDG5UDA>
> .
>

mstv commented:
> fixed that previous change for combined diff did not work

You have caught me: I have used it roundabout twice ever.
I found nothing in the code, will run it tomorrow.
OK to squash.

gerhardol commented:
>>   fixed that previous change for combined diff did not work

> You have caught me: I have used it roundabout twice ever.

Similar for me. Somehow similar to the diff for range-diff, it is interesting to see that there are messages
With fast-forward-only there are no combined diff, still a few non trivial merges on GitHub though
I had seen it when working on this but did not bother fixing it at first, but now I realized I could simplify the code and fix it the same time...

I also want a new button for only showing the range-diff headers, but it is low prio:
https://github.com/gitextensions/gitextensions/pull/8516#discussion_r502685180

gerhardol commented:
@msftbot rebase in 3 days

gerhardol commented:
@msftbot merge when @mstv approves

gerhardol commented:
@msftbot merge in 3 days
